### PR TITLE
Fix support for serving 304 Not Modified service worker script responses

### PR DESCRIPTION
**Pull request by avargaz**
Also fix some comments from the PHPCS ~, according to `WordPress-VIP-Go` and `WordPressVIPMinimum` standards~.

### CODE REVIEW

**Comment by avargaz** [comment]
@googlebot I signed it!

File: wp-includes/class-wp-https-ui.php [hunk 1]
```diff
@@ -370,7 +370,7 @@ public function conditionally_redirect_to_https() {
 			if ( isset( $parsed_url['port'] ) ) {
 				$url .= ':' . $parsed_url['port'];
 			}
-			$url .= wp_unslash( $_SERVER['REQUEST_URI'] );
+			$url .= wp_unslash( sanitize_text_field( $_SERVER['REQUEST_URI'] ) );
```
Comment (westonruter) [inline wp-includes/class-wp-https-ui.php:373]: I don't think this should be done. Core doesn't do it, for example. You can search core for `wp_unslash.+?REQUEST_URI` and find many such examples:

```
src/wp-activate.php:23:list( $activate_path ) = explode( '?', wp_unslash( $_SERVER['REQUEST_URI'] ) );
src/wp-admin/widgets.php:400:					'return' => urlencode( remove_query_arg( wp_removable_query_args(), wp_unslash( $_SERVER['REQUEST_URI'] ) ) ),
src/wp-admin/edit-comments.php:129:	wp_redirect( remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
src/wp-admin/network/user-new.php:85:				$edit_link = esc_url( add_query_arg( 'wp_http_referer', urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ), get_edit_user_link( $user_id_new ) ) );
src/wp-admin/nav-menus.php:622:						'return' => urlencode( remove_query_arg( wp_removable_query_args(), wp_unslash( $_SERVER['REQUEST_URI'] ) ) ),
src/wp-admin/includes/class-wp-users-list-table.php:445:			$edit_link = esc_url( add_query_arg( 'wp_http_referer', urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ), get_edit_user_link( $user_object->ID ) ) );
src/wp-admin/includes/theme.php:651:						'return' => urlencode( esc_url_raw( remove_query_arg( wp_removable_query_args(), wp_unslash( $_SERVER['REQUEST_URI'] ) ) ) ),
src/wp-admin/includes/class-wp-ms-users-list-table.php:273:			$edit_link = esc_url( add_query_arg( 'wp_http_referer', urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ), get_edit_user_link( $user->ID ) ) );
src/wp-admin/includes/class-wp-ms-users-list-table.php:504:			$edit_link       = esc_url( add_query_arg( 'wp_http_referer', urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ), get_edit_user_link( $user->ID ) ) );
src/wp-admin/includes/class-wp-ms-users-list-table.php:509:			$actions['delete'] = '<a href="' . esc_url( network_admin_url( add_query_arg( '_wp_http_referer', urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ), wp_nonce_url( 'users.php', 'deleteuser' ) . '&amp;action=deleteuser&amp;id=' . $user->ID ) ) ) . '" class="delete">' . __( 'Delete' ) . '</a>';
src/wp-admin/edit-tags.php:75:	$referer = wp_unslash( $_SERVER['REQUEST_URI'] );
src/wp-admin/edit-tags.php:212:	$location = remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) );
src/wp-admin/users.php:88:	$referer = '<input type="hidden" name="wp_http_referer" value="' . esc_attr( wp_unslash( $_SERVER['REQUEST_URI'] ) ) . '" />';
src/wp-admin/users.php:455:			wp_redirect( remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
src/wp-admin/users.php:504:									urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ),
src/wp-admin/menu.php:190:	$customize_url            = add_query_arg( 'return', urlencode( remove_query_arg( wp_removable_query_args(), wp_unslash( $_SERVER['REQUEST_URI'] ) ) ), 'customize.php' );
src/wp-admin/upload.php:210:	wp_redirect( remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
src/wp-admin/plugin-install.php:31:	$location = remove_query_arg( '_wp_http_referer', wp_unslash( $_SERVER['REQUEST_URI'] ) );
src/wp-admin/user-new.php:297:				$edit_link = esc_url( add_query_arg( 'wp_http_referer', urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ), get_edit_user_link( $user_id_new ) ) );
src/wp-admin/edit.php:211:	wp_redirect( remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
src/wp-admin/admin.php:50:		wp_redirect( admin_url( 'upgrade.php?_wp_http_referer=' . urlencode( wp_unslash( $_SERVER['REQUEST_URI'] ) ) ) );
src/wp-admin/link-manager.php:43:	wp_redirect( remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
src/wp-includes/functions.php:1838:	$referer_field = '<input type="hidden" name="_wp_http_referer" value="' . esc_attr( wp_unslash( $_SERVER['REQUEST_URI'] ) ) . '" />';
src/wp-includes/functions.php:1863:		$ref = 'previous' == $jump_back_to ? wp_get_referer() : wp_unslash( $_SERVER['REQUEST_URI'] );
src/wp-includes/functions.php:1888:	if ( $ref && wp_unslash( $_SERVER['REQUEST_URI'] ) !== $ref && home_url() . wp_unslash( $_SERVER['REQUEST_URI'] ) !== $ref ) {
src/wp-includes/feed.php:626:	return set_url_scheme( 'http://' . $host['host'] . wp_unslash( $_SERVER['REQUEST_URI'] ) );
src/wp-includes/class-wp-customize-manager.php:2096:		$self_url           = empty( $_SERVER['REQUEST_URI'] ) ? home_url( '/' ) : esc_url_raw( wp_unslash( $_SERVER['REQUEST_URI'] ) );
src/wp-includes/cron.php:678:		wp_redirect( add_query_arg( 'doing_wp_cron', $doing_wp_cron, wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
src/wp-login.php:896:		list( $rp_path ) = explode( '?', wp_unslash( $_SERVER['REQUEST_URI'] ) );
```

So this should have the relevant sanitization rule ignored from PHPCS.

File: wp-includes/class-wp-service-workers.php [hunk 2]
```diff
@@ -152,7 +152,7 @@ public function serve_request() {
 		$file_hash = md5( $output );
 		@header( "ETag: $file_hash" ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.PHP.NoSilencedErrors.Discouraged
 
-		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
+		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( sanitize_text_field( $_SERVER['HTTP_IF_NONE_MATCH'] ) ) : false;
```
Comment (westonruter) [inline wp-includes/class-wp-service-workers.php:155]: This is unnecessary as the value is being used entirely as part of a comparison. And core is not doing this:

```
src/wp-admin/load-styles.php:47:if ( isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) && stripslashes( $_SERVER['HTTP_IF_NONE_MATCH'] ) === $wp_version ) {
src/wp-admin/load-scripts.php:48:if ( isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) && stripslashes( $_SERVER['HTTP_IF_NONE_MATCH'] ) === $wp_version ) {
src/wp-includes/class-wp.php:461:			if ( isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ) {
src/wp-includes/class-wp.php:462:				$client_etag = wp_unslash( $_SERVER['HTTP_IF_NONE_MATCH'] );
src/wp-includes/ms-files.php:65:$client_etag = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? stripslashes( 
 $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
```

What is missing however is `wp_unslash()` and the variable name is wrong:

```suggestion
		$if_none_match = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( wp_unslash( $_SERVER['HTTP_IF_NONE_MATCH'] ) ) : false;
```

File: wp-includes/class-wp-service-workers.php [hunk 3]
```diff
@@ -152,7 +152,7 @@ public function serve_request() {
 		$file_hash = md5( $output );
 		@header( "ETag: $file_hash" ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.PHP.NoSilencedErrors.Discouraged
```
Comment (westonruter) [inline wp-includes/class-wp-service-workers.php:153]: Just noticed this needs to be quoted per [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#Directives):

```suggestion
		$etag      = sprintf( '"%s"', $file_hash );
		@header( "ETag: $etag" ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.PHP.NoSilencedErrors.Discouraged
```

File: wp-includes/class-wp-service-workers.php [hunk 4]
```diff
@@ -152,7 +152,7 @@ public function serve_request() {
...
 
-		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
+		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( sanitize_text_field( $_SERVER['HTTP_IF_NONE_MATCH'] ) ) : false;
 		if ( $file_hash === $etag_header ) {
```
Comment (westonruter) [inline wp-includes/class-wp-service-workers.php:156]:
```suggestion
		if ( $if_none_match === $etag ) {
```

Comment (westonruter) [reply wp-includes/class-wp-service-workers.php:156]: For more on what should be done here, refer to what core is doing: https://github.com/WordPress/wordpress-develop/blob/803eab428d54755acf83403a49bb1cb9c3663c13/src/wp-includes/class-wp.php#L456-L476

**Review by westonruter — CHANGES_REQUESTED** [review]

File: wp-includes/class-wp-https-ui.php [hunk 5]
```diff
@@ -217,7 +217,7 @@ public function render_https_settings() {
 					$url           = $all_insecure_urls[ $i ];
 					$truncated_url = $this->get_truncated_url( $url );
 					?>
-					<li><a href="<?php echo esc_attr( $url ); ?>"><?php echo esc_html( $truncated_url ); ?></a></li>
+					<li><a href="<?php echo esc_url( $url ); ?>"><?php echo esc_html( $truncated_url ); ?></a></li>
```
Comment (westonruter) [inline wp-includes/class-wp-https-ui.php:220]: Good catch on this and the instance below!

File: wp-includes/service-workers.php [hunk 6]
```diff
@@ -225,7 +225,7 @@ function wp_disable_script_concatenation() {
 	 * for browsers that don't support service workers. Note that concatenation only applies in the admin,
 	 * for authenticated users without full-page caching.
 	*/
-	if ( isset( $_COOKIE['wordpress_sw_installed'] ) ) {
+	if ( isset( $_COOKIE['wordpress_sw_installed'] ) ) {  // phpcs:ignore WordPressVIPMinimum.Variables.RestrictedVariables.cache_constraints___COOKIE
```
Comment (westonruter) [inline wp-includes/service-workers.php:228]: The `WordPressVIPMinimum` ruleset is not part of this project, so it doesn't make sense to reference it. If this is causing problems for your install, I suggest ignoring it in a custom ruleset on your end.

```suggestion
	if ( isset( $_COOKIE['wordpress_sw_installed'] ) ) {
```

File: wp-includes/template.php [hunk 7]
```diff
@@ -158,7 +158,7 @@ function wp_service_worker_get_error_messages() {
  */
 function wp_service_worker_error_details_template( $output = '' ) {
 	if ( empty( $output ) ) {
-		$output = '<details id="error-details"><summary>' . esc_html__( 'More Details', 'pwa' ) . '</summary>{{{error_details_iframe}}}</details>';
+		$output = '<details id="error-details"><summary>' . esc_html__( 'More Details', 'pwa' ) . '</summary>{{{error_details_iframe}}}</details>'; // phpcs:ignore WordPressVIPMinimum.Security.Mustache.OutputNotation
```
Comment (westonruter) [inline wp-includes/template.php:161]: Ditto with above in regards to `WordPressVIPMinimum`.

```suggestion
		$output = '<details id="error-details"><summary>' . esc_html__( 'More Details', 'pwa' ) . '</summary>{{{error_details_iframe}}}</details>';
```

**Review by westonruter — CHANGES_REQUESTED** [review]

**Review by westonruter — APPROVED** [review]
Thanks a lot.

**Comment by avargaz** [comment]
Thanks to you @westonruter ! You basically did everything 😄 
Cheers!

**Pull request merged**